### DISCUSSION TITLE
Request for assistance with Navigation in large e-commerce app

### DISCUSSION THREAD

**Discussion by rafaelrcamargo** [react-navigation/react-navigation#11196] [Q&A] [answered]
Hello,

I am currently working on an e-commerce app that has a large number of screens, including product detail pages (PDP), product list pages (PLP), browser authentication sections, and more. I am trying to determine the best way to handle navigation in such a large app, while including a bottom tab navigator and a top search bar.

I would greatly appreciate any guidance or tips on how to effectively handle navigation in this type of app. Any suggestions for libraries or tools that could make this process easier would also be greatly appreciated.

Thank you for your time and help!

Best regards,
Rafael.

**Comment by mitchellkager** [comment, marked as answer]
Hey Rafael, I will offer you the experience I have settled on for my app which has posts, user profiles, a post feed, search functionalities, etc.

I would also encourage you to look at products more similar to what you are creating and imagining how you might mock the navigation in this library.

The core experience of my app spans ~4 functions so I dedicate a tab to each. This is my "core" tab navigator. Many of the tabs have a stack navigator within them (more on that later).

Many of my tabs feature an infinitely scrolling feed. Without going into too much detail, I use `SWR` library to fetch pages and mutate the page items into SWR cache, then pass the item IDs down into pagelets within a`@shopify/Flashlist`. I define a component that will do that fetching using a callback passed as a prop so that I can use it across pages that might need to fetch using different parameters to my API.
The simplest fetch callback is on my homepage which pretty much calls my API with no parameters.
I also have a search page which is essentially the same as the homepage, but gated behind a base stack page that has the user provide filters that are passed to the pushed list view and consequently my API.
[ A previous version of my app embedded the search bar atop the search feed as you describe - I eventually opted for the prerequisite filter screen because there are too many fields on which my content can be filtered for a header bar. Eventually I might migrate to a modal that can be popped open via some interaction in the header. ]
The profile tab is another infinite list with a filter for the auth'd user's profile.

As I mentioned, my app has posts and user profiles. These can be accessed from any of these list views. So, across any tab that has this list view, the tab has to have a stack. I define a helper "shared screen" component that accepts a Navigator (the stack navigator within which these screens will populate) and defines/returns a screen for each component that needs to be accessible from each of these tab stacks.
[ One of my biggest open problems with navigation is that I cannot properly configure identical linking to each of the shared components across my tabs' stacks. ]

For authenticated views: remember that I have a "core" tab navigator? Well, beneath that, my actual root Navigator is a stack that currently only has two screens: my tab navigator, and a transparent modal auth form. If a screen needs to be gated behind authentication, attempting to focus it will instead pop open the auth modal.
[
My current implementation of this auth gating has two problems:
- it uses `onTabPress` to intercept navigation event, which means that user can still navigate past auth gate via programmatic flows... I need to add `focus` listeners
- the intercepted navigation event is never pushed onto the stack, so auth attempt (successful or cancelled) will put the user to the previous page and require re-navigation... Unfortunately, the ideal experience (returning to the requested page on auth success) may require passing params to the auth form so it knows where to return flow back to
]
My auth modal features a stack within itself to handle auth flows: login, registration, confirming 2fa, forgot password, etc.

That is about all of my app that touches this library.

If/when my app defines myriad more functionalities, I will probably have to consider consolidating from (one tab per function) to (one tab per function group) where the latter would use a nested drawer navigator to allow the user navigate between the various functions of the tab's function group.

I also make extensive use of react's built-in context manager to synchronize the state and centralize the logic for:
- an auth context which exposes creds as well as functions that modify auth state (login/out).
- a screen size context that uses a window dimensions hook to provide all descendants with window dimensions as well as a widely utilized calculated sizing
- favicon

Hope this might offer you some direction on how you will structure your app. Good luck!

**Reply by rafaelrcamargo** [reply]
That's amazing, first thanks for your time telling about your app, looks like a pretty solid structure. The infinite scrolling with `Flashlist` is also a thing in my backlog, thanks!
